src/crypto/muhash.cpp

```diff
@@ -0,0 +1,281 @@
+// Copyright (c) 2017 The Bitcoin Core developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#include <crypto/muhash.h>
+
+#include <crypto/chacha20.h>
+#include <crypto/common.h>
+
+#include <assert.h>
+#include <stdio.h>
+
+#include <limits>
+
+namespace {
+
+/** 2^3072 - 1103717, the next largest 3072-bit safe prime number, is used as the order of the group. */
+#define MAX_PRIME_DIFF 1103717
+
+/** Extract the lowest limb of [c0,c1,c2] into n, and left shift the number by 1 limb. */
+#define extract3(c0,c1,c2,n) { \
```

Comment:
Why are these implemented as macros instead of (inline) functions?

Reply:
I am not aware of a reason why inline functions couldn't be used here. From my side, the only reason is that this is @sipa 's original code and I did not feel so strongly about it that it justified a bigger change. Of course, I will do it if that's preferred by reviewers.

Reply:
It may have been that I copied this from code that was originally written for C.

Feel free to change it; inline functions would be far more C++ish.